server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java

```diff
@@ -1180,10 +1208,62 @@ class CacheUnavailableException extends Exception {
    }
 }
 
-class PartialResponse {
-   final Optional<ByteBuf> buffer;
+class OperationContext {
+   final RequestHeader header;
+   final MemcachedParameters parameters;
+   final byte[] key;
+   final byte[] value;
+
+   OperationContext(RequestHeader header, MemcachedParameters parameters, byte[] key, byte[] value) {
+      this.header = Objects.requireNonNull(header);
+      this.parameters = parameters;
+      this.key = key;
+      this.value = value;
+   }
+
+   boolean isNoReply() {
+      return parameters != null && parameters.noReply;
+   }
+
+   boolean isOperation(MemcachedOperation operation) {
+      return header.operation == operation;
+   }
+}
 
-   PartialResponse(Optional<ByteBuf> buffer) {
-      this.buffer = buffer;
+class ResponseEntry implements BiConsumer<Object, Throwable>, Runnable {
+   volatile Object response;
+   volatile Throwable throwable;
+   volatile CompletionStage<Void> responseSent;
+   final MemcachedDecoder decoder;
+   final Channel ch;
+
+   ResponseEntry(MemcachedDecoder decoder, Channel ch) {
+      this.decoder = decoder;
+      this.ch = ch;
+   }
+
+   void queueResponse(CompletionStage<Object> operationResponse) {
+      AggregateCompletionStage<Void> all = CompletionStages.aggregateCompletionStage();
```

Comment:
I am guessing the reduced throughput is because of this method. I think we need to have a lastResponse per channel, so that we are ordering only responses for the same channel and not the entire decoder. The way it is if you have multiple client connections, it will have to wait until another client request is completed before it can proceed. We can store the lastResponse as a channel attribute instead `Channel.attr(AttributeKey.valueOf("MemCachedPreviousResponse");` or something similar.